tests/suite/graceful_recovery_test.go

```diff
@@ -0,0 +1,316 @@
+package suite
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	v1 "k8s.io/api/batch/v1"
+	coordination "k8s.io/api/coordination/v1"
+	core "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+
+	"github.com/nginxinc/nginx-gateway-fabric/tests/framework"
+)
+
+const (
+	teaURL             = "https://cafe.example.com/tea"
+	coffeeURL          = "http://cafe.example.com/coffee"
+	nginxContainerName = "nginx"
+	ngfContainerName   = "nginx-gateway"
+)
+
+var _ = Describe("Graceful Recovery test", Ordered, Label("nfr", "graceful-recovery"), func() {
+	files := []string{
+		"graceful-recovery/cafe.yaml",
+		"graceful-recovery/cafe-secret.yaml",
+		"graceful-recovery/gateway.yaml",
+		"graceful-recovery/cafe-routes.yaml",
+	}
+	ns := &core.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "graceful-recovery",
+		},
+	}
+
+	BeforeEach(func() {
+		Expect(resourceManager.Apply([]client.Object{ns})).To(Succeed())
+		Expect(resourceManager.ApplyFromFiles(files, ns.Name)).To(Succeed())
+		Expect(resourceManager.WaitForAppsToBeReady(ns.Name)).To(Succeed())
+
+		Expect(waitForWorkingTraffic()).ToNot(HaveOccurred())
```

Comment:
I think it's better to use gomega Eventually when we need to wait for something to happen then to use wait.PollUntilContextCance, since we're already heavily using gomega

https://onsi.github.io/gomega/#eventually

It supports specifying durations and timeouts

Here is an example https://github.com/nginxinc/nginx-gateway-fabric/blob/e08292455fd1bf06a8bff1378d076d0d122ba249/tests/suite/telemetry_test.go#L73

This comment applies to other similar places when we call Expect(func()...).ToNot(HaveOccurred()), where func calls wait.PollUntilContextCance

Reply:
Just to understand, you're recommending refactoring to use gomega Eventually **instead** of using wait.PollUntilContextCancel. Meaning I would change something like

```
func waitForWorkingTraffic() error {
	ctx, cancel := context.WithTimeout(context.Background(), timeoutConfig.RequestTimeout)
	defer cancel()

	//nolint:nilerr
	return wait.PollUntilContextCancel(
		ctx,
		500*time.Millisecond,
		true, /* poll immediately */
		func(_ context.Context) (bool, error) {
			if err := expectRequestToSucceed(teaURL, address, "URI: /tea"); err != nil {
				return false, nil
			}
			if err := expectRequestToSucceed(coffeeURL, address, "URI: /coffee"); err != nil {
				return false, nil
			}
			return true, nil
		},
	)
}
```

->

```
func expectWorkingTraffic() error {
  if err := expectRequestToSucceed(teaURL, address, "URI: /tea"); err != nil {
    return false, nil
    }
    if err := expectRequestToSucceed(coffeeURL, address, "URI: /coffee"); err != nil {
    return false, nil
    }
    return true, nil
}

...

using Eventually(expectWorkingTraffic(arguments here...))
```